DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,518 @@
+
+```d
+struct A
+{
+    this(this) {}
+    @implicit this(ref A another) {}        // error, struct A defines a postblit
+}
+```
+
+#### Semantics
+
+The copy constructor typecheck is identical to the constructor one [6][7].
+
+The copy constructor overloads can be explicitly disabled:
+
+```d
+struct A
+{
+    @disable @implicit this(ref A another)
+    @implicit this(ref immutable A another)
+}
+
+void main()
+{
+    A a, b;
+    a = b;     // error: disabled copy construction
+
+    immutable A c;
+    A d = c;  // ok
+
+}
+```
+
+In order to disable copy construction, all copy constructor overloads need to be disabled.
+In the above example, only copies from mutable to mutable are disabled; the overload for
+immutable to mutable copies is still callable.
+
+#### Overloading
+
+The copy constructor can be overloaded with different qualifiers applied to the parameter
+(copying from qualified source) or to the copy constructor itself (copying to qualified
+destination):
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}                        // 1 - mutable source, mutable destination
+    @implicit this(ref immutable A another) {}              // 2 - immutable source, mutable destination
+    @implicit this(ref A another) immutable {}              // 3 - mutable source, immutable destination
+    @implicit this(ref immutable A another) immutable {}    // 4 - immutable source, immutable destination
+}
+
+void main()
+{
+    A a, b, a1;
+    immutable A ia, ib, ia1;
+
+    a = b;      // calls 1
+    a1 = ia;     // calls 2
+    ia = a;     // calls 3
+    ia1 = ib;    // calls 4
+}
+```
+The proposed model enables the user to define the copying from an object of any qualified type
+to an object of any qualified type: any combination of 2 between mutable, `const`, `immutable`, `shared`,
+`const shared`.
+
+The `inout` qualifier may be used for the copy constructor parameter in order to specify that mutable, `const` or `immutable` types are
+treated the same:
+
+```d
+struct A
+{
+    @implicit this(ref inout A another) immutable
+    {
+    }
+}
+
+void main()
+{
+    A a;
+    const A b;
+    immutable A c, r1, r2, r3;
+
+    // All call the same copy construcor because `inout` acts like a wildcard
+    a = r1;
+    b = r2;
+    c = r3;
+}
+```
+
+In case of partial matching, the existing overloading and implicit conversions
+apply to the argument.
+
+#### Interaction with `alias this`
+
+There are situations in which a struct defines both an `alias this` and a copy constructor, and
+for which assignments to variables of the struct type may lead to ambiguities:
+
+```d
+struct A
+{
+    int a;
+    immutable(A) fun()
+    {
+        return immutable A(7);
+    }
+
+    alias fun this;
+
+    @implicit this(ref A another) immutable {}
+}
+
+struct B
+{
+    int a;
+    A fun()
+    {
+        return A(7);
+    }
+
+    alias fun this;
+
+    @implicit this(ref A another) immutable {}
```

Comment:
typo: type should be `ref A`

Reply:
no, it would be an error that way. The type of the copy constructor parameter has to be `typeof(this)`

Reply:
Hmm... I could swear struct B (in code below) had @implicit this(ref A another) immutable {} and I meant to comment on that :) Probably that code was squashed.

Never mind, now it's OK anyway.